src/sentry/api/authentication.py

```diff
@@ -298,6 +299,58 @@ def authenticate(self, request: Request):
 class UserAuthTokenAuthentication(StandardAuthentication):
     token_name = b"bearer"
 
+    def _find_or_update_token_by_hash(self, token_str: str) -> ApiToken | ApiTokenReplica:
+        """
+        Find token by hash or update token's hash value if only found via plaintext.
+        1. Hash provided plaintext token.
+        2. Perform lookup based on hashed value.
+        3. If found, return the token.
+        4. If not found, search for the token based on its plaintext value.
+        5. If found, update the token's hashed value and return the token.
+        6. If not found via hash or plaintext value, raise AuthenticationFailed
+        """
+
+        hashed_token = hashlib.sha256(token_str.encode()).hexdigest()
+
+        if SiloMode.get_current_mode() == SiloMode.REGION:
+            try:
+                # Try to find the token by its hashed value first
+                return ApiTokenReplica.objects.get(hashed_token=hashed_token)
+            except ApiTokenReplica.DoesNotExist:
+                try:
+                    # If we can't find it by hash, use the plaintext string
+                    api_token_replica = ApiTokenReplica.objects.get(token=token_str)
+                except ApiTokenReplica.DoesNotExist:
+                    # If the token does not exist by plaintext either, it is not a valid token
+                    raise AuthenticationFailed("Invalid token")
+                else:
+                    # If found by plaintext, update the ApiToken with the hashed value
+                    # which will replicate back to the region
+                    api_token = ApiToken.objects.get(token=token_str)
+                    api_token.hashed_token = hashed_token
+                    api_token.save(update_fields=["hashed_token"])
+                    return api_token_replica
```

Comment:
Am I able to access `ApiToken` in a region silo to be able to update the `hashed_token` value?

Reply:
Yes... yes this is a problem.